fabric-network/src/impl/query/query.js

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright 2020 IBM All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict';
+
+const logger = require('fabric-network/lib/logger').getLogger('Query');
+
+/**
+ * @typedef {Object} Query~QueryResponse
+ * @memberof module:fabric-network
+ * @property {number} status - The status value from the endorsement. This attriibute
+ * will be set by the chaincode.
+ * @property {Buffer} payload - The payload value from the endorsement. This attribute
+ * may be considered the query value if the status code is exceptable.
+ * @property {Buffer} payload - The message value from the endorsement. This attribute
+ * will have a value when there is not a payload and status value indicates an issue
+ * with determining the payload (the query value).
+ */
+/**
+ * Used by query handler implementations to evaluate transactions on peers of their choosing.
+ * @memberof module:fabric-network
+ */
+class Query {
+	/**
+	 * Builds a Query instance to send and then work with the results returned
+	 * by the fabric-common/Query.
+	 * @param {module:fabric-common.Query} query - The query instance of the proposal
+	 * @returns {Object} options - options to be used when sending the request to
+	 * fabric-common service endpoint {Endorser} peer.
+	 */
+	constructor(query, options = {}) {
+		this.query = query;
+		this.requestTimeout = 3000; // default 3 seconds
+		if (Number.isInteger(options.timeout)) {
+			this.requestTimeout = options.timeout * 1000; // need ms;
+		}
+	}
+
+	/**
+	 * Sends a signed proposal to the specified peers. The peer endorsment
+	 * responses are
+	 * @param {Endorser[]} peers - The peers to query
+	 * @returns {Object.<String, (QueryResponse | Error)>} Object with peer name keys and associated values that are either
+	 * QueryResponse objects or Error objects.
+	 */
+	async evaluate(peers) {
+		const method = 'evaluate';
+		logger.debug('%s - start', method);
+
+		const results = {};
+		try {
+			const responses = await this.query.send({targets: peers, requestTimeout: this.requestTimeout});
+			if (responses) {
+				if (responses.errors) {
+					for (const resultError of responses.errors) {
+						results[resultError.connection.name] = resultError;
+						logger.debug('%s - problem with query to peer %s error:%s', method, resultError.connection.name, resultError);
+					}
+				}
+				if (responses.responses) {
+					for (const peer_response of responses.responses) {
+						if (peer_response.response) {
+							const response = {};
+							response.status = peer_response.response.status;
+							response.payload = peer_response.response.payload;
+							response.message = peer_response.response.message;
+							results[peer_response.connection.name] = response;
```

Comment:
I thought we were going to return the whole `peer_response` so the handler (that may be written by clients) had as much information as possible to make decisions on how to react? Or do you think this is all they should ever need and it's better to keep a layer of abstraction between client code and the peer_response object?

Reply:
So what I am thinking today is the query is completely different, where not getting an endorsement may be the answer they are looking for, not an error. I think that the handler should be presented with:
status : number
payload : bytes
message : string
isEndorsed : boolean
This should provide all the information required to make a decision to return results or try again with a different set of peers.